src/pg/mod.rs

```diff
@@ -724,6 +728,28 @@ impl AsyncPgConnection {
             .unwrap_or_else(|p| p.into_inner())
             .on_connection_event(event);
     }
+
+    pub fn notification_stream(
+        &mut self,
+    ) -> impl futures_core::Stream<Item = diesel::pg::PgNotification> + '_ {
```

Comment:
It's about this signature. I changed it to Stream<Notification> to match the channel, but now I think maybe it's wrong because you said it should be Stream<QueryResult<Notification>>.

That said, the poll_messages function exposes Ok(Notification | Notice) | Err, so if I get an error I can't be sure it's from a notification, a notice, or anything. It's basically just an error.

So it feels to me that I should either output the 3 possible cases, or just pick the one I want (notification). But I'm not sure of this. Anyway, I think diesel sync might have gone through the same question?

Reply:
I think it needs to be `Stream<QueryResult>` as we currently miss one important point: We still need to poll the `drive_connection` future as well as part of the stream, as otherwise we won't get any new notifications while the user polls on the channel we returned. I think that connection future might return an error (via the error channel), that then must be returned from the notification stream.